[PSCE-138] Feat: Add Leveraged Authz

afflom opened the pull request:
This feature enables leveraged authorizations from a leveraged ssp to be added into a leveraging ssp when a user uses the `--leveraged-ssp` flag with ssp assemble. 

My initial approach here is to add a new class to the CRM packages that would contain the logic for extracting the system information from the leveraged ssp's system implementation, converting that into a leveraged authorization, and then adding that leveraged authorization to the leveraging ssp.

afflom committed:
start draft

Signed-off-by: Alex Flom <alexander.flom@gmail.com>

AleJo2995 committed:
fix: parameter aggregation fix (#1443)

* fix: parameter aggregation fix

Signed-off-by: Alejandro Jose Leiva Palomo <alejandro.leiva.palomo@ibm.com>

* fix: arranging tests

Signed-off-by: Alejandro Jose Leiva Palomo <alejandro.leiva.palomo@ibm.com>

* fix: triggering build

Signed-off-by: Alejandro Jose Leiva Palomo <alejandro.leiva.palomo@ibm.com>

* fix: increase time out for cache response

Signed-off-by: Alejandro Jose Leiva Palomo <alejandro.leiva.palomo@ibm.com>

* fix: profile-values are shown in markdown even when there are values already

Signed-off-by: Alejandro Jose Leiva Palomo <alejandro.leiva.palomo@ibm.com>

* fix: adding alt identifier validation

Signed-off-by: Alejandro Jose Leiva Palomo <alejandro.leiva.palomo@ibm.com>

* fix: correct profile values validation

Signed-off-by: Alejandro Jose Leiva Palomo <alejandro.leiva.palomo@ibm.com>

* fix: remove parameter aggregation from assembly and remove label being shown in assembled profile

Signed-off-by: Alejandro Jose Leiva Palomo <alejandro.leiva.palomo@ibm.com>

* fix: correcting test failures and various formatting issues

Signed-off-by: Alejandro Jose Leiva Palomo <alejandro.leiva.palomo@ibm.com>

* fix: change order for parameters in markdown

Signed-off-by: Alejandro Jose Leiva Palomo <alejandro.leiva.palomo@ibm.com>

* fix: not setting empty values

Signed-off-by: Alejandro Jose Leiva Palomo <alejandro.leiva.palomo@ibm.com>

---------

Signed-off-by: Alejandro Jose Leiva Palomo <alejandro.leiva.palomo@ibm.com>

jpower432 committed:
feat: updates ssp-generate to filter control implementation for leveraged_ssp

Signed-off-by: Jennifer Power <barnabei.jennifer@gmail.com>

jpower432 committed:
refactor: updates ExportWriter to reduce code duplication

Signed-off-by: Jennifer Power <barnabei.jennifer@gmail.com>

jpower432 committed:
fix: updates ExportReader to add new statements if present in the inheritance view

Signed-off-by: Jennifer Power <barnabei.jennifer@gmail.com>

afflom committed:
saving get_system

Signed-off-by: Alex Flom <alexander.flom@gmail.com>

jpower432 committed:
fix: update logging to debug in ExportReader

Signed-off-by: Jennifer Power <barnabei.jennifer@gmail.com>

BE-Code committed:
fix: fixing typo in encoding name (#1448)

Co-authored-by: AleJo2995 <alejandro.leiva.palomo@ibm.com>

afflom committed:
works. saving.

Signed-off-by: Alex Flom <alexander.flom@gmail.com>

jpower432 commented on tests/trestle/core/crm/leverager_test.py:
```diff
@@ -0,0 +1,68 @@
+        assert isinstance(result, ossp.SystemComponent)
+'''
+
+def test_leverager_init(sample_leveraging_ssp, sample_leveraged_components, sample_trestle_root):
```
If you look under here - https://github.com/RedHatProductSecurity/compliance-trestle/blob/15578ec86746dc9c547ae631ffdf3e07b9276dc2/tests/conftest.py#L208. There is already a test fixture for creating a trestle working directory.

jpower432 commented on trestle/core/crm/leverager.py:
```diff
@@ -0,0 +1,111 @@
+
+    def _create_leveraging_components(self) -> List[ossp.SystemComponent]:
+        """Create leveraging components."""
+        self._leveraging_components = []
```
It looks like you are altering this field twice. You are append to `self._leveraged_components`, but then you are also setting the field to itself in the constructor on line 57.

jpower432 commented on trestle/core/crm/leverager.py:
```diff
@@ -0,0 +1,111 @@
+        href: pathlib.Path = ModelUtils.get_model_path_for_name_and_class(trestle_root, leveraged_ssp, ossp.SystemSecurityPlan) # type: ignore
+        if href is None:
+            logger.error("Failed to get model path.")
+            return leveraged_authz# Handle the error appropriately, perhaps by raising an exception or returning
```
You can raise this error - https://github.com/RedHatProductSecurity/compliance-trestle/blob/266f67bd220e15922caacb4de0e702f4d0927ceb/trestle/common/err.py#L23. Perhaps it would make more sense to look this up in the CLI/commands layer for quicker feedback and just pass in the href information as an argument.

jpower432 commented on trestle/core/crm/leverager.py:
```diff
@@ -0,0 +1,111 @@
+        logger.debug("Creating leveraged authorizations.")
+        leveraged_authz = gens.generate_sample_model(ossp.LeveragedAuthorization)
+        # Insert the relative leveraged ssp path as a link
+        leveraged_authz.links = []
```
Trestle has some list_utils that could simplify this a bit. Example of how it is used here - https://github.com/IBM/compliance-trestle/blob/bdf60b26075f7250bcdbbe08745630b27042ad74/trestle/core/control_interface.py#L1131

jpower432 committed:
Merge branch 'develop' into feat/add-inheritance-view

jpower432 committed:
refactor: simplify code in read_exports_from_markdown

Signed-off-by: Jennifer Power <barnabei.jennifer@gmail.com>

jpower432 committed:
feat: adds method to ExportReader to track leveraged_component from Markdown

Signed-off-by: Jennifer Power <barnabei.jennifer@gmail.com>

afflom committed:
saving for filtering

Signed-off-by: Alex Flom <alexander.flom@gmail.com>

jpower432 committed:
Merge remote-tracking branch 'aflom/feat/add-leveraged-authz' into PSCE-220

Commit:
Merge pull request #1 from jpower432/PSCE-220

PSCE-220

jpower432 committed:
feat: Adds SSPInheritanceAPI for consolidated logic

Updates markdown to writed leveraged SSP href to markdown instead
of retrieving as a flag from assembly

Signed-off-by: Jennifer Power <barnabei.jennifer@gmail.com>

jpower432 committed:
tests: adds addititonal tests for SSPInheritanceAPI

Adds common inheritance_file test generation

Signed-off-by: Jennifer Power <barnabei.jennifer@gmail.com>

jpower432 committed:
test: adds test in update ssp to check for idempotent operations

Signed-off-by: Jennifer Power <barnabei.jennifer@gmail.com>

Commit:
Merge pull request #2 from jpower432/PSCE-235

PSCE-235
